### PR TITLE
refactor(lang-rust): Setting up Rust environment properly

### LINKED ISSUE #464: rustup update not working
@aledbf you asked me to open a separate issue for this, so here we go:

```
gitpod /workspace/XXX $ rustc --version
rustc 1.52.1 (9bc8c42bb 2021-05-09)

gitpod /workspace/XXX $ rustup update
info: no updatable toolchains installed
error: rustup is not installed at '/workspace/.cargo'
```

currently the container has an older version of rust installed and so I thought the regular workflow to update rust inside my gitpod would work, but it does not.

### PR DESCRIPTION
**Pull request by axonasif**
## Description
- About #464
> I came across three solutions
> 1. Wrapping `rustup` with a bash script to unset `CARGO_HOME` variable to make it use the original `$HOME/.cargo` dir. Might not be future proof. (Current proposed workaround)
> 2. Greedily symlinking everything from `$HOME/.cargo/bin` to `/workspace/.cargo/bin`(Custom `CARGO_HOME`). Might cause unexpected issues.
> 3. Using shell specific aliases to wrap `rustup` in a quite neat way, ~~what I ended up with.~~

- About #506 
> Previously this:
> https://github.com/gitpod-io/workspace-images/blob/7d640a13abc662a459f885f7a432e8d9337d9a4e/chunks/lang-rust/Dockerfile#L27
> was being run with having custom `CARGO_HOME`(**/workspace/.cargo**). But `/workspace` is non-existent in the actual container image since `/workspace` is basically mounted when a gitpod session is running, so any previous content under `/workspace` will not be effective due to the mount.

## Related Issue(s)

Fixes #464 #506 https://github.com/gitpod-io/workspace-images/issues/642

## How to test

- Spin up a custom Gitpod workspace based on this image.
- You can also run this image inside an workspace as per the gitpod docs and test partially since there will be no `/workspace` inside the container.
- Run: `ls $HOME/.cargo/bin` to check if additional cargo submodules are installed properly  for #506 
- Run: `rustup update` for #464

## Release Notes

```release-note
`rustup update` should work now
Cargo submodules in the `lang-rust` image should be installed correctly
```

## Documentation

No change required.

### CODE REVIEW

**Review by gengjiawen — APPROVED** [review]
Thx

**Comment by gengjiawen** [comment]
@jankeromnes Can you take a look. I verified this, it works.

**Comment by jankeromnes** [comment]
Many thanks @gengjiawen!

@gitpod-io/engineering-workspace please take a look 🙏

**Comment by axonasif** [comment]
> Many thanks @gengjiawen!
> 
> @gitpod-io/engineering-workspace please take a look 🙏

Hi @jankeromnes, `@gitpod-io/engineering-workspace` is ending up in a `404` page for me 😅

**Comment by jankeromnes** [comment]
Aha, sorry for the confusion there 😅 yes indeed, I was pinging Gitpod's workspace team, asking them to take a look at your changes when they have some time. I hope this works! 🤞

**Comment by gengjiawen** [comment]
> Aha, sorry for the confusion there 😅 yes indeed, I was pinging Gitpod's Workspace team, asking them to take a look at your changes when they have some time. I hope this works! 🤞

Looks this repo really lack of reviewing those days. Is there any way to speed this up ?

**Comment by axonasif** [comment]
> Looks this repo really lack of reviewing those days. Is there any way to speed this up ?

You can check this tweet 😀
https://twitter.com/mikenikles/status/1471485721366073352?s=19

**Comment by gengjiawen** [comment]
> > Looks this repo really lack of reviewing those days. Is there any way to speed this up ?
> 
> You can check this tweet 😀 [twitter.com/mikenikles/status/1471485721366073352?s=19](https://twitter.com/mikenikles/status/1471485721366073352?s=19)

lol, this picture is real art.

**Comment by gengjiawen** [comment]
@jankeromnes Happy new year and any update from the team ?

**Comment by aledbf** [comment]
@gengjiawen  thank you for your contribution. At this moment, we are revamping the build process of the images.
For that reason, we are not introducing new features/images or changes. Once we finish, we will review your pull request.
Apologies for the inconvenience.

**Comment by gengjiawen** [comment]
> @gengjiawen at this moment, we are revamping the build process of the images. For that reason, we are not introducing new features/images or changes. Once we finish, we will review your pull request. Apologies for the inconvenience.
> 
> Edit: please check #593

Is this can be done in this week ? I also have two PRs pending.

**Comment by gengjiawen** [comment]
@axonasif Can you rebase this ? looks the refactor is merged.

**Comment by axonasif** [comment]
> @axonasif Can you rebase this ? looks the refactor is merged.

Thanks for the heads up ❤️

File: chunks/lang-rust/Dockerfile [hunk 1]
```diff
@@ -6,22 +6,22 @@ USER gitpod
...
-# TODO: setting CARGO_HOME to /workspace/.cargo avoids manual updates. Remove after full workspace backups are GA.
+    && rustup completions bash | sudo tee /etc/bash_completion.d/rustup.bash-completion > /dev/null \
+    && rustup completions bash cargo | sudo tee /etc/bash_completion.d/rustup.cargo-bash-completion > /dev/null \
+    && printf '%s\n' "$(grep -v -F -x -f /home/gitpod/.profile_orig /home/gitpod/.profile)" \
+    'mkdir -m 0755 -p "${CARGO_HOME:-/workspace/.cargo}" 2>/dev/null' > $HOME/.bashrc.d/80-rust \
+    && mkdir -m 0755 -p $HOME/.config/fish \
+    && printf '%s\n' 'alias rustup="env -u CARGO_HOME (command -v rustup)"' >> $HOME/.config/fish/config.fish \
+    && printf '%s\n' 'alias rustup="env -u CARGO_HOME $(command -v rustup)"' >> $HOME/.bashrc.d/80-rust \
+    && printf '%s\n' 'alias rustup="env -u CARGO_HOME $(command -v rustup)"' >> $HOME/.zshrc
+RUN cargo install cargo-watch cargo-edit cargo-tree cargo-workspaces
```
Comment (baptiste0928) [inline chunks/lang-rust/Dockerfile:25]:
```suggestion
RUN cargo install cargo-watch cargo-edit cargo-workspaces
```

`cargo-tree` has been integrated in Cargo since 1.44 and [its repository](https://github.com/sfackler/cargo-tree) is archived.

File: legacy/full/Dockerfile [hunk 2]
```diff
@@ -183,26 +183,26 @@ ENV PATH=/workspace/.rvm/bin:$PATH
...
-# TODO: setting CARGO_HOME to /workspace/.cargo avoids manual updates. Remove after full workspace backups are GA.
+    && rustup completions bash | sudo tee /etc/bash_completion.d/rustup.bash-completion > /dev/null \
+    && rustup completions bash cargo | sudo tee /etc/bash_completion.d/rustup.cargo-bash-completion > /dev/null \
+    && printf '%s\n' "$(grep -v -F -x -f /home/gitpod/.profile_orig /home/gitpod/.profile)" \
+    'mkdir -m 0755 -p "${CARGO_HOME:-/workspace/.cargo}" 2>/dev/null' > $HOME/.bashrc.d/80-rust \
+    && mkdir -m 0755 -p $HOME/.config/fish \
+    && printf '%s\n' 'alias rustup="env -u CARGO_HOME (command -v rustup)"' >> $HOME/.config/fish/config.fish \
+    && printf '%s\n' 'alias rustup="env -u CARGO_HOME $(command -v rustup)"' >> $HOME/.bashrc.d/80-rust \
+    && printf '%s\n' 'alias rustup="env -u CARGO_HOME $(command -v rustup)"' >> $HOME/.zshrc
+RUN cargo install cargo-watch cargo-edit cargo-tree cargo-workspaces
```
Comment (baptiste0928) [inline legacy/full/Dockerfile:202]:
```suggestion
RUN cargo install cargo-watch cargo-edit cargo-workspaces
```

See previous comment.

**Review by baptiste0928 — COMMENTED** [review]
Thanks for fixing the Rust installation :grinning: 

I don't know what are the selection criteria of installed cargo plugins, but I would suggest adding some of [the most popular ones](https://crates.io/categories/development-tools::cargo-plugins?sort=downloads):
- [`cargo-audit`](https://lib.rs/crates/cargo-audit): check known security vulnerabilities in crates listed in `Cargo.lock`
- [`cargo-outdated`](https://lib.rs/crates/cargo-outdated): check for outdated dependencies
- [`cargo-deny`](https://lib.rs/crates/cargo-deny): lint dependencies (security advisories, licenses, ...)
- [`cargo-make`](https://lib.rs/crates/cargo-make): task runner and build tool

As I said in https://github.com/gitpod-io/workspace-images/issues/642, I also think that launching `cargo watch` in the default configuration in redundant as rust-analyzer (installed by default on Rust workspaces) already regularly run `cargo check`.

**Comment by princerachit** [comment]
Hey @axonasif can you update this PR so it fixes new dazzlev2 images rather?

File: legacy/full/Dockerfile (hunk 2 above)
Comment (axonasif) [reply legacy/full/Dockerfile:202]: I guess I will no longer update the legacy image.

**Comment by axonasif** [comment]
> Hey @axonasif can you update this PR so it fixes new dazzlev2 images rather?

I think I already did, or am I missing something 😅 ?

**Comment by axonasif** [comment]
@baptiste0928 
> I don't know what are the selection criteria of installed cargo plugins, but I would suggest adding some of the [most popular](https://crates.io/categories/development-tools::cargo-plugins?sort=downloads) (and useful) ones:
> 
> * [`cargo-audit`](https://lib.rs/crates/cargo-audit): check known security vulnerabilities in crates listed in `Cargo.lock`
> * [`cargo-outdated`](https://lib.rs/crates/cargo-outdated): check for outdated dependencies
> * [`cargo-deny`](https://lib.rs/crates/cargo-deny): lint dependencies (security advisories, licenses, ...)
> * [`cargo-make`](https://lib.rs/crates/cargo-make): task runner and build tool

We certainly could but then there would be no point in using the `minimal` installation profile on the image, I'm not the one who initially wrote the rust setup for gitpod here. So can't really provide an argument why but I will try to stick with the topic of this PR, maybe that's something for another day 😄 

> As I said in #642, I also think that launching `cargo watch` in the default configuration in redundant as rust-analyzer (installed by default on Rust workspaces) already regularly run `cargo check`.

Currently, we have to keep it this way until the `.gitpod.yml` autogen code for rust projects is synced up.

**Comment by baptiste0928** [comment]
Thanks for your answer, should I create a new issue for these suggestions?

File: chunks/lang-rust/Dockerfile [hunk 3]
```diff
@@ -6,22 +6,22 @@ USER gitpod
...
-    && .cargo/bin/rustup completions bash cargo | sudo tee /etc/bash_completion.d/rustup.cargo-bash-completion > /dev/null \
-    && grep -v -F -x -f /home/gitpod/.profile_orig /home/gitpod/.profile > /home/gitpod/.bashrc.d/80-rust
-ENV PATH=$HOME/.cargo/bin:$PATH
-# TODO: setting CARGO_HOME to /workspace/.cargo avoids manual updates. Remove after full workspace backups are GA.
+    && rustup completions bash | sudo tee /etc/bash_completion.d/rustup.bash-completion > /dev/null \
+    && rustup completions bash cargo | sudo tee /etc/bash_completion.d/rustup.cargo-bash-completion > /dev/null \
+    && printf '%s\n' "$(grep -v -F -x -f /home/gitpod/.profile_orig /home/gitpod/.profile)" \
+    'mkdir -m 0755 -p "${CARGO_HOME:-/workspace/.cargo}" 2>/dev/null' > $HOME/.bashrc.d/80-rust \
+    && mkdir -m 0755 -p $HOME/.config/fish \
+    && printf '%s\n' 'alias rustup="env -u CARGO_HOME (command -v rustup)"' >> $HOME/.config/fish/config.fish \
```
Comment (princerachit) [inline chunks/lang-rust/Dockerfile:22]: missing `$` ? in `(command -v rustup)`?

File: legacy/full/Dockerfile [hunk 4]
```diff
@@ -183,26 +183,26 @@ ENV PATH=/workspace/.rvm/bin:$PATH
 LABEL dazzle/layer=lang-rust
```
Comment (princerachit) [inline legacy/full/Dockerfile:183]: Please revert these changes. We are not updating the legacy images anymore.

File: chunks/lang-rust/Dockerfile [hunk 5]
```diff
@@ -6,22 +6,22 @@ USER gitpod
...
+    && printf '%s\n' 'alias rustup="env -u CARGO_HOME $(command -v rustup)"' >> $HOME/.bashrc.d/80-rust \
+    && printf '%s\n' 'alias rustup="env -u CARGO_HOME $(command -v rustup)"' >> $HOME/.zshrc
```
Comment (princerachit) [inline chunks/lang-rust/Dockerfile:24]: I am not sure about this. Are we updating other shells in other languages? I doubt it. Let us stick to the pattern. You can refer to contribution guide to learn why we prefer creating a dedicated rc in rc.d rather than appending it to existing rc files.

**Review by princerachit — CHANGES_REQUESTED** [review]

File: chunks/lang-rust/Dockerfile (hunk 3 above)
Comment (princerachit) [reply chunks/lang-rust/Dockerfile:22]: Please write tests for the changes. Again, you can refer to contribution guide to learn how to test these changes, I will also encourage you to build the image locally once and test it.

Comment (axonasif) [reply chunks/lang-rust/Dockerfile:22]: That's the `fish` syntax. I and @gengjiawen had manually test this image before the dazzleV2 rewrite.

Comment (axonasif) [reply chunks/lang-rust/Dockerfile:22]: Will test again now 🙏

File: legacy/full/Dockerfile (hunk 4 above)
Comment (axonasif) [reply legacy/full/Dockerfile:183]: Oh yeah, missed this part. Thank you for spotting 🙏

File: chunks/lang-rust/Dockerfile (hunk 5 above)
Comment (axonasif) [reply chunks/lang-rust/Dockerfile:24]: This is my proposed solution, please check the PR description. Re-posting a part of the desc here:

- About #464
> I came across three solutions
> 1. Wrapping `rustup` with a bash script to unset `CARGO_HOME` variable to make it use the original `$HOME/.cargo` dir. Might not be future proof.
> 2. Greedily symlinking everything from `$HOME/.cargo/bin` to `/workspace/.cargo/bin`(Custom `CARGO_HOME`). Might cause unexpected issues.
> 3. Using shell specific aliases to wrap `rustup` in a quite neat way, what I ended up with.

> @princerachit: You can refer to contribution guide to learn why we prefer creating a dedicated rc in rc.d rather than appending it to existing rc files.

Yeah I totally agree, I'm doing that for the `bash` one(`$HOME/.bashrc.d/80-rust`), but for zsh and fish we don't have any custom logic set in their main shellrc file for automatically loading dedicated rc.d files, which is why I'm directly appending into the main shellrc. That's something we need to do from the `base` image I suppose.

**Comment by axonasif** [comment]
> @baptiste0928 Thanks for your answer, should I create a new issue for these suggestions?

Yeah, please go ahead!

File: chunks/lang-rust/Dockerfile (hunk 5 above)
Comment (princerachit) [reply chunks/lang-rust/Dockerfile:24]: >which is why I'm directly appending into the main shellrc. That's something we need to do from the base image I suppose.

I agree. That is why I would prefer not to have the shell changes as part of the PR. This will set a wrong expectation for future contributors. We need to figure out if we want to support other shells. Can you create an issue for it?

Comment (princerachit) [reply chunks/lang-rust/Dockerfile:24]: Since we were not supporting other shells let us skip it for now. We can revisit it later if required

Comment (axonasif) [reply chunks/lang-rust/Dockerfile:24]: I actually made the attempt to support other shells in this specific PR after finding out that the `base` image installs `zsh` and `fish` alongside `bash` 😅 
https://github.com/gitpod-io/workspace-images/blob/d43c719bb2ff7b6849c0456fe21e3ca06c20168f/base/Dockerfile#L30

Comment (axonasif) [reply chunks/lang-rust/Dockerfile:24]: Will create an issue and let you know 🙏

Comment (axonasif) [reply chunks/lang-rust/Dockerfile]: I went back to my proposed solution 1 instead of using shell aliases.

File: chunks/lang-rust/Dockerfile (hunk 3 above)
Comment (axonasif) [reply chunks/lang-rust/Dockerfile]: Ran the build and did manual tests too, all good so far.

File: legacy/full/Dockerfile [hunk 6]
```diff
@@ -184,7 +184,7 @@ LABEL dazzle/layer=lang-rust
 LABEL dazzle/test=tests/lang-rust.yaml
 USER gitpod
 RUN cp /home/gitpod/.profile /home/gitpod/.profile_orig && \
-    curl -fsSL https://sh.rustup.rs | sh -s -- -y --profile minimal --default-toolchain 1.58.0 \
+    curl -fsSL https://sh.rustup.rs | sh -s -- -y --profile minimal --default-toolchain 1.56.1 \
```
Comment (axonasif) [inline legacy/full/Dockerfile:187]: This was on behalf of the commit revert, just FYI.

**Comment by axonasif** [comment]
Hey @princerachit, can you please review again 🙏 ?

File: chunks/lang-rust/Dockerfile [hunk 7]
```diff
@@ -6,22 +6,26 @@ USER gitpod
 # Dazzle does not rebuild a layer until one of its lines are changed. Increase this counter to rebuild this layer.
 ENV TRIGGER_REBUILD=1
 
-RUN cp /home/gitpod/.profile /home/gitpod/.profile_orig && \
-    curl -fsSL https://sh.rustup.rs | sh -s -- -y --profile minimal --default-toolchain 1.58.1 \
-    && .cargo/bin/rustup component add \
+ENV PATH=$PATH:$HOME/.cargo/bin
```
Comment (princerachit) [inline chunks/lang-rust/Dockerfile:9]: Read more [here](https://github.com/gitpod-io/workspace-images/blob/master/CONTRIBUTING.md#creating-new-image):
>Adding a chunk -> Here is a list of best practices you should keep in mind while adding a chunk: -> 5. Always add new path as prefix to existing path. e.g. ENV PATH=/my-tool/path/bin:$PATH. Not doing so can cause path conflicts and can potentially break other images.

File: chunks/lang-rust/Dockerfile [hunk 8]
```diff
@@ -6,22 +6,26 @@ USER gitpod
...
-# TODO: setting CARGO_HOME to /workspace/.cargo avoids manual updates. Remove after full workspace backups are GA.
+    && rustup completions bash | sudo tee /etc/bash_completion.d/rustup.bash-completion > /dev/null \
+    && rustup completions bash cargo | sudo tee /etc/bash_completion.d/rustup.cargo-bash-completion > /dev/null \
+    && printf '%s\n' "$(grep -v -F -x -f /home/gitpod/.profile_orig /home/gitpod/.profile)" \
+                        'mkdir -m 0755 -p "${CARGO_HOME:-/workspace/.cargo}" 2>/dev/null' > $HOME/.bashrc.d/80-rust \
+    && _rustup_path="$(command -v rustup)" && mv "$_rustup_path" "${_rustup_path}.main" \
+    && printf '%s\n' '#!/usr/bin/bash -eu' \
+                        'exec env -u CARGO_HOME "$(command -v rustup.main)" "$@"' > "$_rustup_path" \
+    && chmod 0755 "$_rustup_path"
+    # && mkdir -m 0755 -p $HOME/.config/fish \
```
Comment (princerachit) [inline chunks/lang-rust/Dockerfile:25]: If you have an issue for this then simply add it in the comment or remove the entire commented block.

File: legacy/full/Dockerfile (hunk 6 above)
Comment (princerachit) [reply legacy/full/Dockerfile:187]: There should be absolutely no change in the `/legacy` dir. Please rebase/squash your commits so there is no changes in this PR wrt to legacy dir

**Review by princerachit — CHANGES_REQUESTED** [review]

File: chunks/lang-rust/Dockerfile [hunk 9]
```diff
@@ -6,22 +6,26 @@ USER gitpod
...
+    # && printf '%s\n' 'alias rustup="env -u CARGO_HOME (command -v rustup)"' >> $HOME/.config/fish/config.fish \
+    # && printf '%s\n' 'alias rustup="env -u CARGO_HOME $(command -v rustup)"' >> $HOME/.bashrc.d/80-rust \
+    # && printf '%s\n' 'alias rustup="env -u CARGO_HOME $(command -v rustup)"' >> $HOME/.zshrc
+RUN cargo install cargo-watch cargo-edit cargo-workspaces
```
Comment (princerachit) [inline chunks/lang-rust/Dockerfile:29]: can we add tests for these?

**Comment by princerachit** [comment]
Is the description upto date?

File: chunks/lang-rust/Dockerfile (hunk 8 above)
Comment (axonasif) [reply chunks/lang-rust/Dockerfile:25]: Okay got it, thanks for pointing out 🙏

File: legacy/full/Dockerfile (hunk 6 above)
Comment (axonasif) [reply legacy/full/Dockerfile:187]: Got it, my bad, there might have been some conflict. Will revert that change.

**Comment by axonasif** [comment]
> Is the description upto date?

I had to update the line references to dockerfiles 👍

File: chunks/lang-rust/Dockerfile (hunk 9 above)
Comment (axonasif) [reply chunks/lang-rust/Dockerfile:26]: I'm afraid, I wasn't able to make the new tests pass in the test-env for cargo submodules. It works when the image is externally run by docker or used in `.gitpod.yml image:`

Comment (axonasif) [reply chunks/lang-rust/Dockerfile:26]: And it's also difficult to test the changes on `tests/lang-rust.yml` since `chunks/lang-rust/Dockerfile` is re-built each time `test/lang-rust.yml` was modified, it almost takes 5 mins each time and sometimes gets stuck after a long period of time in the workspace.

File: tests/lang-rust.yaml [hunk 10]
```diff
@@ -2,24 +2,44 @@
...
   entrypoint: [bash, -i, -c]
   assert:
-  - status == 0
+    - status == 0
 # - desc: it should have matching rustc and cargo versions
 #   command: [cargo --version; rustc --version]
 #   entrypoint: [bash, -i, -c]
 #   assert:
 #   - status == 0
 #   - stdout.split("\n")[0].split(" ")[1] == stdout.split("\n")[1].split(" ")[1]
```
Comment (princerachit) [inline tests/lang-rust.yaml:25]: This will not work because stdout can be read only once. AFAIK atm dazzle does not allow you to create a variable out of stdout/stderr.

File: tests/lang-rust.yaml [hunk 11]
```diff
@@ -2,24 +2,44 @@
   command: [rustc --version]
```
Comment (princerachit) [inline tests/lang-rust.yaml:2]: Here is a working patch with some fixed tests:
```git
diff --git a/tests/lang-rust.yaml b/tests/lang-rust.yaml
index e846162..596e224 100644
--- a/tests/lang-rust.yaml
+++ b/tests/lang-rust.yaml
@@ -2,44 +2,39 @@
   command: [rustc --version]
   entrypoint: [bash, -i, -c]
   assert:
-    - status == 0
-    - stdout.indexOf("rustc") != -1
+  - status == 0
+  - stdout.indexOf("rustc 1.58") != -1
 - desc: it should have properly configured rustup
   command: [rustup show]
   entrypoint: [bash, -i, -c]
   assert:
-    - status == 0
-    - stdout.indexOf("/home/gitpod/.rustup") != -1
-    - stdout.indexOf("x86_64-unknown-linux-gnu") != -1
-    - stdout.indexOf("x86_64-unknown-linux-musl") == -1
+  - status == 0
+  - stdout.indexOf("/home/gitpod/.rustup") != -1
+  - stdout.indexOf("x86_64-unknown-linux-gnu") != -1
+  - stdout.indexOf("x86_64-unknown-linux-musl") == -1
 - desc: it should have cargo
   command: [cargo --version]
   entrypoint: [bash, -i, -c]
   assert:
-    - status == 0
-# - desc: it should have matching rustc and cargo versions
-#   command: [cargo --version; rustc --version]
-#   entrypoint: [bash, -i, -c]
-#   assert:
-#   - status == 0
-#   - stdout.split("\n")[0].split(" ")[1] == stdout.split("\n")[1].split(" ")[1]
+  - status == 0
+  - stdout.indexOf("cargo 1.58") != -1
 
 ## The below doesn't work in tests env but works in actual image shell, had to give up
-# - desc: it should have watch subcommand for cargo
-#   entrypoint: [bash, -i, -c]
-#   command: [cargo watch --help]
-#   assert:
-#     - status == 0
-#     - stdout.indexOf("cargo-watch") != -1
-# - desc: it should have one of edit(add, rm, upgrade, set-version) subcommand(s) for cargo
-#   entrypoint: [bash, -i, -c]
-#   command: [cargo, add, --help]
-#   assert:
-#     - status == 0
-#     - stdout.indexOf("cargo-add") != -1
-# - desc: it should have workspaces subcommand for cargo
-#   entrypoint: [bash, -i, -c]
-#   command: [cargo, workspaces, --help]
-#   assert:
-#     - status == 0
-#     - stdout.indexOf("cargo-workspaces") != -1
+- desc: it should have watch subcommand for cargo
+  entrypoint: [bash, -i, -c]
+  command: [cargo watch --help]
+  assert:
+  - status == 0
+  - stdout.indexOf("cargo-watch") != -1
+- desc: it should have one of edit(add, rm, upgrade, set-version) subcommand(s) for cargo
+  entrypoint: [bash, -i, -c]
+  command: [cargo add --help]
+  assert:
+  - status == 0
+  - stdout.indexOf("cargo-add") != -1
+- desc: it should have workspaces subcommand for cargo
+  entrypoint: [bash, -i, -c]
+  command: [cargo workspaces --help]
+  assert:
+  - status == 0
+  - stdout.indexOf("cargo-workspaces") != -1
```

Comment (princerachit) [reply tests/lang-rust.yaml:2]: Please fix the alignments and spacing.

File: tests/lang-rust.yaml (hunk 10 above)
Comment (axonasif) [reply tests/lang-rust.yaml:25]: Umm, I didn't write that code block of `it should have matching rustc and cargo versions`.

File: tests/lang-rust.yaml (hunk 11 above)
Comment (axonasif) [reply tests/lang-rust.yaml:2]: Okay so I was able to make those three tests pass by masking the `CARGO_HOME` command which points to `/workspace/.cargo`. `cargo` is not able to locate the subcommand binaries when that's set for **dazzle test env**, surprisingly it isn't an issue when I run the image with `docker run` and try something like `cargo add --help`, I'm not sure why this happens and how to debug it for dazzle tests. It might be because that `/workspace` dir is non-existent in the resulting docker images/containers but oddly it doesn't affect `docker run localhost:5000/dazzle:full`

Comment (axonasif) [reply tests/lang-rust.yaml:2]: Also, it's very difficult to go and manually fix all the indentations as VSCODE auto indents the lines, but guess I'll do it anyways 😅 
If this is very important, I think we should disable auto indentation for this repo by using `.vscode/settings.json` to make the contribution flow easier.

Comment (axonasif) [reply tests/lang-rust.yaml:2]: Sorry, I can't fix the indents from VSCODE on Gitpod 😓

File: tests/lang-rust.yaml [hunk 12]
```diff
@@ -2,24 +2,42 @@
...
   entrypoint: [bash, -i, -c]
   assert:
-  - status == 0
-  - stdout.indexOf("rustc") != -1
+    - status == 0
```
Comment (axonasif) [inline tests/lang-rust.yaml:5]: Sorry, I can't fix the indents from VSCODE on Gitpod 😓
